testbed/sampling/sampling_data_provider.go

```diff
@@ -0,0 +1,176 @@
+package sampling
+
+import (
+	"encoding/binary"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// samplingDataProvider is an implementation of DataProvider for use in trace tests
+// This data provider will generator some traces that fall within the provided option set
+// and some data that does not. This is useful for testing filter and sampling processors.
+type samplingDataProvider struct {
+	// Used by the load generator to report how many itens were created
+	dataItemsGenerated *atomic.Uint64
+	// Number of traces to generate in a run
+	numTracesToGenerate int
+	// Number of spans per trace to generate in a run
+	spansPerTrace   int
+	traceIDSequence atomic.Uint64
+	// Total traces generated
+	tracesGenerated int
+	// we will use this to ensure that the received spans matches the expected sampled spans
+	sampledSpansGenerated atomic.Uint64
+	// Controls if a trace should be sampled
+	shouldSample shouldSampleFunc
+	// Customize spans that need to be sampled so that they can properly filtered
+	sampledSpanCustomizer sampledSpanCustomizerFunc
+	// Stores all the trances that were sampled
+	expectedSampledTraces []ptrace.Traces
+}
+
+type sampledSpanCustomizerFunc func(span ptrace.Span)
+
+type shouldSampleFunc func(traceIDSequence uint64) bool
+
+// SamplingDataProviderOption defines a PollutedDataProvider option.
+type SamplingDataProviderOption func(t *samplingDataProvider)
+
+// WithShouldSampleFunction allows you to control if a trace should be sampled.
+func WithSamplingFunction(shouldSample shouldSampleFunc) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.shouldSample = shouldSample
+	}
```

Comment:
This is the only thing I don't fully understand the usefulness of. Does this need to be so generic? Given a traceid what other ways can you generate a sampling decision? In your example you use `traceid%10 < 3` to generate what I think is intended to be < 30 percent. 

Are we better off passing in some type of hardcoded value and worrying about the implementation of that logic? What other sampling decisions do you envision being implemented with this generality? To me it was a bit confusing on how I was supposed to implement this optional function.

Also, should this be optional or required?

Reply:
your questions make totally sense. I updated the code to remove options that are not currently used in our tests. we can more generic behaviour once it is really needed. otherwise the defaults for the tests works.